File: plugins/ble-peripheral/AliveBlePeripheralModule.kt
/**
 * AliveBlePeripheralModule — Android BLE GATT 서버 + 광고 네이티브 모듈
 *
 * BluetoothGattServer로 SERVICE_UUID 서비스를 제공하고,
 * BluetoothLeAdvertiser로 해당 서비스를 광고합니다.
 * JS에서 startPeripheral/stopPeripheral로 제어합니다.
 */

package com.xrx.aliveconnection

import android.bluetooth.*
import android.bluetooth.le.*
import android.content.Context
import android.os.ParcelUuid
import com.facebook.react.bridge.*
import java.util.UUID

class AliveBlePeripheralModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String = "AliveBlePeripheral"

    private var gattServer: BluetoothGattServer? = null
    private var advertiser: BluetoothLeAdvertiser? = null
    private var isRunning = false

    /**
     * BLE Peripheral(GATT 서버 + 광고) 시작
     * @param serviceUuid 서비스 UUID 문자열
     * @param charUuid userId characteristic UUID 문자열
     * @param userId 광고할 유저 ID
     */
    @ReactMethod
    fun startPeripheral(serviceUuid: String, charUuid: String, userId: String, promise: Promise) {
        try {
            val bluetoothManager = reactApplicationContext
                .getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager
            val adapter = bluetoothManager?.adapter

            if (adapter == null || !adapter.isEnabled) {
                promise.reject("BLE_DISABLED", "블루투스가 꺼져 있습니다")
                return
            }

            val sUuid = UUID.fromString(serviceUuid)
            val cUuid = UUID.fromString(charUuid)

            // 1. GATT 서버 생성
            val gattCallback = object : BluetoothGattServerCallback() {
                override fun onCharacteristicReadRequest(
                    device: BluetoothDevice?,
                    requestId: Int,
                    offset: Int,
                    characteristic: BluetoothGattCharacteristic?
                ) {
                    // userId를 UTF-8 바이트로 응답
                    val value = userId.toByteArray(Charsets.UTF_8)
                    gattServer?.sendResponse(
                        device,
                        requestId,
                        BluetoothGatt.GATT_SUCCESS,
                        offset,
                        if (offset > 0) value.copyOfRange(offset, value.size) else value
                    )
                }

                override fun onConnectionStateChange(
                    device: BluetoothDevice?,
                    status: Int,
                    newState: Int
                ) {
                    // 연결/해제 로깅 (필요시 JS 이벤트 발행 가능)
                }
            }

            gattServer = bluetoothManager.openGattServer(reactApplicationContext, gattCallback)

            // 서비스 + characteristic 등록
            val service = BluetoothGattService(sUuid, BluetoothGattService.SERVICE_TYPE_PRIMARY)
            val characteristic = BluetoothGattCharacteristic(
                cUuid,
                BluetoothGattCharacteristic.PROPERTY_READ,
                BluetoothGattCharacteristic.PERMISSION_READ
            )
            characteristic.value = userId.toByteArray(Charsets.UTF_8)
            service.addCharacteristic(characteristic)
            gattServer?.addService(service)

            // 2. BLE 광고 시작
            advertiser = adapter.bluetoothLeAdvertiser
            if (advertiser == null) {
                promise.reject("BLE_NO_ADVERTISER", "이 기기는 BLE 광고를 지원하지 않습니다")
                return
            }

            val settings = AdvertiseSettings.Builder()
                .setAdvertiseMode(AdvertiseSettings.ADVERTISE_MODE_LOW_LATENCY)
                .setTxPowerLevel(AdvertiseSettings.ADVERTISE_TX_POWER_MEDIUM)
                .setConnectable(true)
                .build()

            val data = AdvertiseData.Builder()
                .setIncludeDeviceName(false)
                .addServiceUuid(ParcelUuid(sUuid))
                .build()

            advertiser?.startAdvertising(settings, data, object : AdvertiseCallback() {
                override fun onStartSuccess(settingsInEffect: AdvertiseSettings?) {
                    isRunning = true
                    promise.resolve(true)
                }

                override fun onStartFailure(errorCode: Int) {
                    promise.reject(
                        "BLE_ADV_FAIL",
                        "BLE 광고 시작 실패 (코드: $errorCode)"
                    )
                }
            })
        } catch (e: SecurityException) {
            promise.reject("BLE_PERMISSION", "BLE 권한이 없습니다: ${e.message}")
        } catch (e: Exception) {
            promise.reject("BLE_ERROR", "BLE Peripheral 시작 실패: ${e.message}")
        }
    }

    /**
     * BLE Peripheral(GATT 서버 + 광고) 중지
     */
    @ReactMethod
    fun stopPeripheral(promise: Promise) {
        try {
            advertiser?.stopAdvertising(object : AdvertiseCallback() {})
            gattServer?.clearServices()
            gattServer?.close()
            gattServer = null
            advertiser = null
            isRunning = false
            promise.resolve(true)
        } catch (e: SecurityException) {
            promise.reject("BLE_PERMISSION", "BLE 권한이 없습니다: ${e.message}")
        } catch (e: Exception) {
            promise.reject("BLE_ERROR", "BLE Peripheral 중지 실패: ${e.message}")
        }
    }

    /**
     * 이 기기에서 BLE Peripheral 지원 여부 확인
     */
    @ReactMethod
    fun isSupported(promise: Promise) {
        try {
            val bluetoothManager = reactApplicationContext
                .getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager
            val adapter = bluetoothManager?.adapter
            val supported = adapter?.isMultipleAdvertisementSupported == true
            promise.resolve(supported)
        } catch (e: Exception) {
            promise.resolve(false)
        }
    }
}

File: plugins/ble-peripheral/AliveBlePeripheralModule.swift
/**
 * AliveBlePeripheralModule — iOS BLE Peripheral (CBPeripheralManager) 네이티브 모듈
 *
 * CBPeripheralManager로 GATT 서비스를 제공하고 BLE 광고를 수행합니다.
 * JS에서 startPeripheral/stopPeripheral/isSupported로 제어합니다.
 */

import Foundation
import CoreBluetooth
import React

@objc(AliveBlePeripheral)
class AliveBlePeripheralModule: NSObject, CBPeripheralManagerDelegate {

  private var peripheralManager: CBPeripheralManager?
  private var userId: String = ""
  private var serviceUuid: CBUUID?
  private var charUuid: CBUUID?

  // Promise 콜백 저장 (비동기 초기화 완료 후 resolve)
  private var startResolve: RCTPromiseResolveBlock?
  private var startReject: RCTPromiseRejectBlock?
  private var pendingService: CBMutableService?

  // MARK: - React Native 메서드

  @objc
  func startPeripheral(
    _ serviceUuidStr: String,
    charUuid charUuidStr: String,
    userId userIdStr: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    self.userId = userIdStr
    self.serviceUuid = CBUUID(string: serviceUuidStr)
    self.charUuid = CBUUID(string: charUuidStr)
    self.startResolve = resolve
    self.startReject = reject

    // CBPeripheralManager 생성 (상태 변경 시 delegate 호출)
    peripheralManager = CBPeripheralManager(delegate: self, queue: nil)
  }

  @objc
  func stopPeripheral(
    _ resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    peripheralManager?.stopAdvertising()
    peripheralManager?.removeAllServices()
    peripheralManager = nil
    resolve(true)
  }

  @objc
  func isSupported(
    _ resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    // iOS는 모든 기기가 BLE Peripheral 지원
    resolve(true)
  }

  // MARK: - CBPeripheralManagerDelegate

  func peripheralManagerDidUpdateState(_ peripheral: CBPeripheralManager) {
    switch peripheral.state {
    case .poweredOn:
      setupService()
    case .poweredOff:
      startReject?("BLE_DISABLED", "블루투스가 꺼져 있습니다", nil)
      startResolve = nil
      startReject = nil
    case .unauthorized:
      startReject?("BLE_UNAUTHORIZED", "BLE 권한이 없습니다", nil)
      startResolve = nil
      startReject = nil
    default:
      break
    }
  }

  func peripheralManager(
    _ peripheral: CBPeripheralManager,
    didAdd service: CBService,
    error: Error?
  ) {
    if let error = error {
      startReject?("BLE_SERVICE_ERROR", "서비스 추가 실패: \(error.localizedDescription)", error)
      startResolve = nil
      startReject = nil
      return
    }

    // 서비스 등록 성공 → 광고 시작
    guard let sUuid = serviceUuid else { return }
    peripheral.startAdvertising([
      CBAdvertisementDataServiceUUIDsKey: [sUuid],
      CBAdvertisementDataLocalNameKey: "ALIVE"
    ])
  }

  func peripheralManagerDidStartAdvertising(
    _ peripheral: CBPeripheralManager,
    error: Error?
  ) {
    if let error = error {
      startReject?("BLE_ADV_FAIL", "BLE 광고 시작 실패: \(error.localizedDescription)", error)
    } else {
      startResolve?(true)
    }
    startResolve = nil
    startReject = nil
  }

  func peripheralManager(
    _ peripheral: CBPeripheralManager,
    didReceiveRead request: CBATTRequest
  ) {
    // userId를 UTF-8 바이트로 응답
    guard let cUuid = charUuid, request.characteristic.uuid == cUuid else {
      peripheral.respond(to: request, withResult: .attributeNotFound)
      return
    }

    let data = userId.data(using: .utf8) ?? Data()
    if request.offset > data.count {
      peripheral.respond(to: request, withResult: .invalidOffset)
      return
    }

    request.value = data.subdata(in: request.offset..<data.count)
    peripheral.respond(to: request, withResult: .success)
  }

  // MARK: - Private

  private func setupService() {
    guard let sUuid = serviceUuid, let cUuid = charUuid else { return }

    let characteristic = CBMutableCharacteristic(
      type: cUuid,
      properties: .read,
      value: nil, // nil이면 동적 값 — didReceiveRead에서 응답
      permissions: .readable
    )

    let service = CBMutableService(type: sUuid, primary: true)
    service.characteristics = [characteristic]

    peripheralManager?.add(service)
  }

  // MARK: - React Native 모듈 설정

  @objc
  static func requiresMainQueueSetup() -> Bool {
    return false
  }
}
